draco/synthesis/mockcatalog.py

```diff
@@ -795,19 +809,213 @@ def process(self, cat):
+            cat_z_err[:] = (z_err ** 2 + cat_z_err ** 2) ** 0.5
+
+        return cat
+
+
+class AddEBOSSZErrorsToCatalog(task.SingleTask, random.RandomTask):
+    """Add eBOSS-type random redshift errors to redshifts in a catalog.
+
+    See the docstrings for _{qso,elg,lrg}_velocity_error() for descriptions
+    of each redshift error distribution.
+
+    Note also that the errors are added to the catalog in place, such that
+    if the original catalog is subsequently used in a pipeline, it will
+    have the errors included.
+
+    Attributes
+    ----------
+    tracer : {"ELG"|"LRG"|"QSO"}
+        Generate redshift errors corresponding to this eBOSS sample.
+    """
+
+    tracer = config.enum(["QSO", "ELG", "LRG"])
+
+    def process(self, cat):
+        """Generate random redshift errors and add to redshifts in catalog.
+
+        Parameters
+        ----------
+        cat : :class:`containers.SpectroscopicCatalog`
+            Input catalog.
+
+        Returns
+        ----------
+        cat_out : :class:`containers.SpectroscopicCatalog`
+            Catalog with redshift errors added.
+        """
+
+        # Get redshifts from catalog
+        cat_z = cat["redshift"]["z"][:]
+        cat_z_err = cat["redshift"]["z_error"][:]
+
+        # Generate redshift errors for the chosen tracer
+        z_err = self._generate_z_errors(cat_z, self.tracer)
 
         # Add errors to catalog redshifts
         cat_z += z_err
         # Add errors in quadrature with existing stored errors
-        cat_z_err = (z_err ** 2 + cat_z_err ** 2) ** 0.5
+        cat_z_err[:] = (z_err ** 2 + cat_z_err ** 2) ** 0.5
 
         return cat
 
+    def _generate_z_errors(self, z, tracer):
+        """Generate redshift errors using a tracer-specific velocity error distribution.
+
+        See e.g. Eq. (A1) from https://arxiv.org/abs/1012.2912 for the
+        relationship between redshift errors and peculiar velocity errors.
+
+        Parameters
+        ----------
+        z: np.ndarray[nsource,]
+            Source redshifts.
+        tracer : {"ELG"|"LRG"|"QSO"}
+            Name of the tracer.
+
+        Returns
+        -------
+        dz: np.ndarray[nsource,]
+            Perturbations to source redshifts based on random velocity errors.
+        """
+
+        _velocity_error_function_lookup = {
+            "QSO": self._qso_velocity_error,
+            "ELG": self._elg_velocity_error,
+            "LRG": self._lrg_velocity_error,
+        }
+
+        if tracer not in _velocity_error_function_lookup:
+            raise ValueError(
+                f"Do not recognize {tracer}.  Must define a method "
+                "for drawing random velocity errors for this tracer."
+            )
+
+        err_func = _velocity_error_function_lookup[tracer]
+
+        dv = err_func(z.size, self.rng)
+
+        dz = (1.0 + z) * dv / (C * 1e-3)
+
+        return dz
+
+    @staticmethod
+    def _qso_velocity_error(nsample, rng):
```

Comment:
This all looks great. My only comment would be to remove the `_` prefixes from these so they're no longer "private" and it's clearer that they could be used elsewhere.

Reply:
Changed!